ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoiceNotifier.java

```diff
@@ -132,32 +128,71 @@ private void internalTerminalBlockReached(Bytes32 executionBlockHash) {
-      // Payload ID builds on the correct execution block
-      if (lastPayloadId.isPresent() || parentExecutionHash.isZero()) {
-        return lastPayloadId;
+    final boolean lastForkChoiceStateCorrectlyBuildsOnTop =
+        lastSentForkChoiceState
+            .map(ForkChoiceState::getHeadBlockHash)
+            .map(
+                fcsHead -> {
+                  if (fcsHead.equals(parentExecutionHash)) {
+                    return true;
+                  }
+                  return executionPayloadTerminalBlockHash.isPresent()
+                      && fcsHead.equals(executionPayloadTerminalBlockHash.get());
+                })
+            .orElse(false);
+
+    if (lastForkChoiceStateCorrectlyBuildsOnTop) {
+      // current Future Payload ID builds on the correct block
+
+      if (lastFuturePayloadId.isEmpty()) {
+        // consistency check
+        throw new IllegalStateException(
+            "we cannot have lastSentForkChoiceState without lastFuturePayloadId");
       }
-      throw new IllegalStateException(
-          String.format("PayloadId not available for Beacon Block Root %s", parentBeaconBlockRoot));
+
+      return lastFuturePayloadId
+          .get()
+          .thenApply(
+              payloadId -> {
+
+                // we accept empty payload only in pre-merge
+                if (payloadId.isPresent() || parentExecutionHash.isZero()) return payloadId;
+
+                throw new IllegalStateException(
+                    String.format(
+                        "PayloadId not available for Beacon Block Root %s", parentBeaconBlockRoot));
+              });
     }
+
+    // we have no SentForkChoiceState or doesn't build on top of the right block
     if (parentExecutionHash.isZero()) {
-      // Haven't reached the merge block yet, parent has default payload
-      if (executionPayloadTerminalBlockHash.isPresent()
-          && lastSentForkChoiceHead.equals(executionPayloadTerminalBlockHash.get())) {
-        // Creating the merge block - payload ID is building on the terminal PoW block
-        return lastPayloadId;
+      // pre-merge
+
+      if (executionPayloadTerminalBlockHash.isEmpty()) {
+        // we are in pre-merge and terminal block is not reached, we can build a block with empty
+        // payload
+        return SafeFuture.completedFuture(Optional.empty());
+      }
+
+      if (allowPayloadIdOnTheFlyRetrieval) {
+        // we are in pre-merge and terminal block is reached
+        // try to obtain a payloadId now
+        final Bytes32 terminalBlockHash = executionPayloadTerminalBlockHash.get();
+        final ForkChoiceState terminalForkChoiceState =
+            new ForkChoiceState(terminalBlockHash, terminalBlockHash, Bytes32.ZERO);
+        internalForkChoiceUpdated(terminalForkChoiceState);
+        if (!lastSentForkChoiceState.equals(forkChoiceState) || lastFuturePayloadId.isEmpty()) {
+          throw new IllegalStateException();
```

Comment:
I'm not sure I'd bother with this check.  There would have to be something seriously wrong with the fork choice updated method to not have set this.  If you want to keep it I'd use `Preconditions.checkState` so it looks more like a sanity check rather than important logic.